### ISSUE TITLE
Support for Loss Functions (Symmetric Loss function)

### ISSUE THREAD

**Issue by s0nicboOm** [numaproj/numalogic#71]
# Summary

What change needs making?

# Use Cases

When would you use this?

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by ab93** [comment]
Can you please explain why we would need this? @s0nicboOm

**Comment by s0nicboOm** [comment]
> Can you please explain why we would need this? @s0nicboOm

Symmetric Loss function is when we want to weight the positive as well as the negative outliers equally. Currently we use Absolute error and MSE. 

Here there is an Opportunity of introducing more such similar loss functions.

**Comment by haripriyajk** [comment]
Hey,I would like to try resolving this issue.

**Comment by rum1887** [comment]
hi, If I have understood it right, you want to me add support for symmetric functions https://github.com/numaproj/numalogic/blob/main/numalogic/models/vae/base.py here?

```
def _init_criterion(loss_fn: str) -> Callable:
    if loss_fn == "huber":
        return F.huber_loss
    if loss_fn == "l1":
        return F.l1_loss
    if loss_fn == "mse":
        return F.mse_loss
    raise ValueError(f"Unsupported loss function provided: {loss_fn}")
```

like a sigmoid loss etc?
correct me if I'm wrong 

@vigith @s0nicboOm

**Comment by rum1887** [comment]
torch.nn.functional has 21 loss functions among which 6 are symmetric loss functions, 3 are already supported 

1. L1 Loss (l1_loss)
2. Mean Squared Error Loss (mse_loss)
3. Huber Loss (huber_loss)

Need to add the following
1. Poisson Negative Log Likelihood Loss (poisson_nll_loss)
2. Gaussian Negative Log Likelihood Loss (gaussian_nll_loss)
3. Smooth L1 Loss (smooth_l1_loss)

**Comment by rum1887** [comment]
```
def _init_criterion(loss_fn: str) -> Callable:
    if loss_fn == "huber":
        return F.huber_loss
    if loss_fn == "l1":
        return F.l1_loss
    if loss_fn == "mse":
        return F.mse_loss
if loss_fn == "pnl":
        return F.poisson_nll_loss
if loss_fn == "gnlll":
        return F.gaussian_nll_loss
if loss_fn == "smoothloss":
        return F.smooth_l1_loss
    raise ValueError(f"Unsupported loss function provided: {loss_fn}")
```

**Comment by s0nicboOm** [comment]
> torch.nn.functional has 21 loss functions among which 6 are symmetric loss functions, 3 are already supported
> 
> 1. L1 Loss (l1_loss)
> 2. Mean Squared Error Loss (mse_loss)
> 3. Huber Loss (huber_loss)
> 
> Need to add the following
> 
> 1. Poisson Negative Log Likelihood Loss (poisson_nll_loss)
> 2. Gaussian Negative Log Likelihood Loss (gaussian_nll_loss)
> 3. Smooth L1 Loss (smooth_l1_loss)

Hi @rum1887 !

We are ultimately looking to support loss functions that are not only in PyTorch but also provide a flexibility to the user to plug in there own custom loss function. This issue does not only point to adding more loss functions but also asks for a better way of providing that interface to the user to bring in their own custom loss function that integrates with the models that we have today.

Thanks for the question. Let me add it to the issue summary for better clarification.
